WordPress/Classes/Utility/DiskCache.swift

```diff
@@ -0,0 +1,236 @@
+    }
+
+    // MARK: - Sweep
+
+    private func performSweepIfNeeded() {
+        let sweepDateKey = "disk-cache-last-sweep-date-\(rootURL)"
+        if let sweepDate = UserDefaults.standard.value(forKey: sweepDateKey) as? Date,
+           Date().timeIntervalSince(sweepDate) < sweepInterval {
+            return // The last sweep was completed recently
+        }
+        // Perform the sweep after a brief delay to reduce the pressure on the system
+        // during the app launch
+        queue.asyncAfter(deadline: .now() + .seconds(10)) { [weak self] in
+            guard let self = self else { return }
+            do {
+                try self.sweep()
+                UserDefaults.standard.set(Date(), forKey: sweepDateKey)
+            } catch {
+                DDLogError("Failed to perform cache sweep with error: \(error)")
+            }
+        }
+    }
+
+    func sweep() throws {
+        var entries = try contents(withKeys: [.contentAccessDateKey, .totalFileAllocatedSizeKey])
+        guard !entries.isEmpty else { return }
+        var totalSize = entries.reduce(0) {
+            $0 + ($1.attributes.totalFileAllocatedSize ?? 0)
+        }
+        guard totalSize > sizeLimit else {
+            return // The size is OK
+        }
+        // Removes most entities, but not all (keep 50% of the size limit).
+        let targetSizeLimit = Int(Double(sizeLimit) * 0.5)
+        let distantPath = Date.distantPast // Should never be needed
+        entries.sort { // Most recently accessed items first
+            ($0.attributes.contentAccessDate ?? distantPath) > ($1.attributes.contentAccessDate ?? distantPath)
+        }
+        while totalSize > targetSizeLimit, let entry = entries.popLast() {
+            totalSize -= (entry.attributes.totalFileAllocatedSize ?? 0)
+            try FileManager.default.removeItem(at: entry.url)
+        }
+    }
+
+    private struct FileEntry {
+        let url: URL
+        let attributes: URLResourceValues
+    }
+
+    private func contents(withKeys keys: Set<URLResourceKey> = []) throws -> [FileEntry] {
+        let keys = keys.union([.isRegularFileKey])
+        guard let enumerator = FileManager.default.enumerator(
+            at: rootURL,
+            includingPropertiesForKeys: Array(keys),
+            options: [.skipsHiddenFiles, .skipsPackageDescendants]
+        ) else {
+            return []
+        }
+        var files: [FileEntry] = []
+        for case let fileURL as URL in enumerator {
+            do {
+                let attributes = try fileURL.resourceValues(forKeys: keys)
+                if attributes.isRegularFile ?? false {
+                    files.append(FileEntry(url: fileURL, attributes: attributes))
+                }
+            } catch { print(error, fileURL) }
+        }
+        return files
+    }
+}
+
+private extension String {
+    var sha256: String? {
+        guard !isEmpty, let data = self.data(using: .utf8) else {
+            return nil
```

Comment:
We can produce a sha256 hash for an empty string: 

```
$ printf '' | sha256sum
e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855  -
```

Reply:
The goal was to prevent empty keys from being used by accident. It's a common mistake – happens all the time.

I moved this check here to make it more clear:

```swift
private func fileURL(for key: String) -> URL? {
        guard !key.isEmpty, let filename = key.sha256 else { return nil }
```

Reply:
If it's a mistake, that should not happen. We could use a `precondition` here? Again, by turning this unexpected scenario into an "error case" instead of failing silently, we can let the developers know they have misused the API and should correct their code.

Reply:
Great point. It seems like a perfect scenario for an assertion: we don't want to fail silently in the debug mode, but it should be ignored in production. I added the assertion. Lmkwyt.

Reply:
I'm still not sure what the additional `sha256` overhead does for us here – I feel like we have good tools to avoid empty keys?

One other suggestion, if it helps – make the `key` something like a `NonEmptyString` type (ideally that's `ExpressibleByStringLiteral`, so it'd be super ergonomic) – it's _effectively_ a compile-time correctness check, because it pushes the possibility for invalid keys out of this class and up into its callers.

Reply:
You can't use some characters in filenames, for example `/`.

```swift
 do {
    let cacheKey = "https://wp.com/some-request"
    let fileURL = URL.cachesDirectory
        .appendingPathComponent(cacheKey, isDirectory: false)
    try Data([1, 2, 3]).write(to: fileURL)
} catch {
    print(error)
}
```

Prints:

```
Error Domain=NSCocoaErrorDomain Code=4 "The file “some-request” doesn’t exist." UserInfo={NSFilePath=/Users/kean/Library/Caches/https://wp.com/some-request, NSUnderlyingError=0x600000c63fc0 {Error Domain=NSPOSIXErrorDomain Code=2 "No such file or directory"}}
```

The filename length also has a limit, which is different in different file systems. I think it's 255 character in AFPS.

Reply:
> IMHO it'd make more sense to cache the 3 pages of stats we just fetched, in which case I'd imagine the cache key would be something like ${siteID}-view-stats?

I used manually constructed keys like this in [the PR](https://github.com/wordpress-mobile/WordPress-iOS/pull/21067/files) for which this cache was added. But it's common to use URLs or keys that you have no control over. There needs to be some way to handle it.

I've benchmarked the hashing in the past, and it was negligible compared to the I/O speed. I would not worry about the overhead.

P.S. The URL caching should typically happen on the networking layer and use HTTP caching, which is not what's currently happening as far as I can tell.

> nothing beats being able to look at the file system and see your cache item

Absolutely, but it's probably less of a priority for a cache where, again, you should not rely on it being available. It'd be more of a priority if it was a database.

Speaking about the databases, my preferred solution would've been to use a database as a key-value store for these small JSONs. It's supposed to be [more efficient](https://www.sqlite.org/fasterthanfs.html) both in terms of speed and space, but not a significant difference, and it doesn't help with larger files. In a database, a key can contain any characters and have any length. I decided against using the database, because I wanted to implement something simple enough to get it merged quickly and processed with addressing other issues 😆 FileManager is also very fast, is thread-safe, and is extremely reliable.

`URLCache` uses a database to store (small) responses. I've also looked into [swift-corelibs-foundation](https://github.com/apple/swift-corelibs-foundation/tree/main), and they, surprisingly, are **not** using a database. This implementation also [uses base64](https://github.com/apple/swift-corelibs-foundation/blob/adf511f1bdd5c1a205c5238c166ecd667ef8a60a/Sources/FoundationNetworking/URLCache.swift#L354) to encode the keys. There is an advantage to it base64 is reversible. But I can't see where it handles the filename length limit.

Reply:
So, just to summarize, it may not be the best solution to use hashing, but it has clear advantages, and based on other systems I saw, including the ones in other languages, it's the most common one. I would've also kept SHA1, but it'll be deprecated in 2030 😃